src/rhash/hash.c

```diff
@@ -1223,7 +1255,83 @@ static int rc_hash_wii(md5_state_t* md5,
+static int rc_hash_wiiware(md5_state_t* md5,
+                           const rc_hash_iterator_t* iterator,
+                           void* file_handle)
+{
+  uint32_t cert_chain_size, ticket_size, tmd_size;
+  uint32_t tmd_start_addr, content_count, content_addr, content_size;
+  uint32_t ix;
+
+  uint8_t quad_buffer[4];
+  uint8_t* buffer;
+
+  rc_file_seek(iterator, file_handle, 0x08, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  cert_chain_size =
+    (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  /* Round up to nearest 0x40 */
+  cert_chain_size = (cert_chain_size + 0x3F) & ~0x3F;
+  rc_file_seek(iterator, file_handle, 0x10, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  ticket_size =
+    (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  ticket_size = (ticket_size + 0x3F) & ~0x3F;
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  tmd_size =
+    (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  tmd_size = (tmd_size + 0x3F) & ~0x3F;
+
+  tmd_start_addr = 0x40 + cert_chain_size + ticket_size;
+  if (tmd_size > MAX_BUFFER_SIZE)
+    tmd_size = MAX_BUFFER_SIZE;
+  buffer = (uint8_t*)malloc(tmd_size);
+
+  /* Hash TMD */
+  rc_file_seek(iterator, file_handle, tmd_start_addr, SEEK_SET);
+  rc_file_read(iterator, file_handle, buffer, tmd_size);
+  rc_hash_iterator_verbose_formatted(iterator, "Hashing %u byte TMD", tmd_size);
+  md5_append(md5, buffer, tmd_size);
+  free(buffer);
+
+  /* Get count of content sections */
+  rc_file_seek(iterator, file_handle, (uint64_t)tmd_start_addr + 0x1de, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 2);
+  content_count = (quad_buffer[0] << 8) | quad_buffer[1];
+  rc_hash_iterator_verbose_formatted(iterator, "Hashing %u content sections", content_count);
+  content_addr = tmd_start_addr + tmd_size;
+  for (ix = 0; ix < content_count; ix++)
+  {
+    /* Get content section size */
+    rc_file_seek(iterator, file_handle, (uint64_t)tmd_start_addr + 0x1e4 + 8 + ix * 0x20, SEEK_SET);
+    rc_file_read(iterator, file_handle, quad_buffer, 4);
+    if (quad_buffer[0] == 0x00 && quad_buffer[1] == 0x00 && quad_buffer[2] == 0x00 && quad_buffer[3] == 0x00)
+    {
+      rc_file_read(iterator, file_handle, quad_buffer, 4);
+      content_size =
+        (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+      content_size = (content_size + 0x0F) & ~0x0F;
+      if (content_size > MAX_BUFFER_SIZE)
+        content_size = MAX_BUFFER_SIZE;
+    }
+    else
+    {
+      content_size = MAX_BUFFER_SIZE;
+    }
+    /* Hash content */
+    buffer = (uint8_t*)malloc(content_size);
+    rc_file_seek(iterator, file_handle, content_addr, SEEK_SET);
+    rc_file_read(iterator, file_handle, buffer, content_size);
+    md5_append(md5, buffer, content_size);
+    content_addr += content_size;
```

Comment:
You'll want to store the original `content_size` somewhere. If it happens to be greater than `MAX_BUFFER_SIZE`, you'll only advance by `MAX_BUFFER_SIZE`.